### DISCUSSION TITLE
Looking for TypeScript reviewers

### DISCUSSION THREAD

**Discussion by eliandoran** [TriliumNext/Trilium#6063] [General]
Hello, @TriliumNext/everyone .

As I have already started a bit working on porting the backend to TypeScript, I need TypeScript guys to review my code.

If you have experience with TypeScript, write a comment here and I will create a team containing only you that we can assign for reviews (and potentially development later on).

**Comment by zerebos** [comment]
I will try to take a look when I have some time, but my TypeScript skills are not the best, I generally stick to more simple types (according to TypeScript experts) since it does the task at hand.

**Comment by ProfDoof** [comment]
I'm not an expert on typescript in particular but I'm fairly comfortable with most type systems at this point and am more than happy to look at it. I also have a few friends who are typescript gurus that I can ping to get second opinions on things.

**Comment by eliandoran** [comment]
@rauenzi , @ProfDoof , I would appreciate a review on the conversion PRs I have started: #13, #18, #25

**Comment by eliandoran** [comment]
@TriliumNext/everyone , could anyone else help with reviews for the TypeScript conversion PRs, please?
i would really like to move forward with the TypeScript conversion.

**Reply by Micosilent** [reply]
I left some comments in the #25 PR, I did not leave a proper code review, but did look at all the code and found it good.

**Comment by meichthys** [comment]
Also See #157

**Comment by eliandoran** [comment]
@TriliumNext/development , I am reopening a bit this discussion since it's an important topic for me.

Personally I feel that development is currently going a bit faster than the reviews. I had a few PRs that were unreviewed for a few days. I know that people are busy, but we should really decide on approach that does not reduce efficiency or cause any potential frustrations.

For some PRs this has caused me to merge some PRs without review, since they were needed for the upcoming release or no consensus was reached.

Maybe it's an over-expectation from me since I am accustomed to next-day reviews maximum as part of my job.

cc. @root-hal9000 , @rauenzi

**Reply by JYC333** [reply]
I can leave some comments for your PRs. And I see some PRs failed on some checks, are those checks need to be fixed before reviewing?

**Reply by eliandoran** [reply]
@JYC333 , I would appreciate it.

You can review any PR that is [not in draft](https://github.com/TriliumNext/Notes/pulls?q=is%3Apr+is%3Aopen+-is%3Adraft). #242 and #259 are failing checks due to the large number of modifications (it's converting the server from CJS to ESM), there will be a third PR which fixes them.

**Reply by JYC333** [reply]
Ok, I'll take a look then. And I'm not a member of development team, not sure whether my review would be valid.

**Reply by eliandoran** [reply]
@JYC333 , if you're interested in joining, let me know.

Any review is valid provided you have experience with JS/TS.

**Reply by JYC333** [reply]
Yes, I'd like to join. I have develop some ts plugins and websites by my own. Not an expert, but at least have some experience.

**Reply by eliandoran** [reply]
@JYC333 , done. Feel free to join our Matrix as well (if you're not already part of it).

**Reply by zerebos** [reply]
Yeah I definitely get where you're coming from Elian, but unfortunately I have a lot going on outside of Trilium to take care of and cannot spend all my time on it.

> Maybe it's an over-expectation from me since I am accustomed to next-day reviews maximum as part of my job.

This is how it works at my job too generally, but that's because we are all spending 8 hours a day at our jobs. This is not at all how it works in open-source in my experience. Things tend to move slow in the review phase especially for larger changes, it feels like an exponential increase.

> For some PRs this has caused me to merge some PRs without review, since they were needed for the upcoming release or no consensus was reached.

In my opinion since we are making such large sweeping changes anyways, I am not opposed to seeing development go fast like this without extensive reviews. Only when we have solidifies everything and have a userbase do we really need truly extensive reviews. I think right now as long as the changes are something everyone is on the same page about then it should be fine to move fast.

**Reply by eliandoran** [reply]
@rauenzi , Indeed open-source appears to be a different kind of beast. 🤔 

I was thinking maybe it's a better idea to go with a similar approach as @zadam , which is simply to take lead on the development and see where that gets us.

Regarding "as the changes are something everyone is on the same page", I think we'd have a hard time aligning everyone on everything. And then comes the philosophical question: if there's only a handful of developers, why should they cater to everyone's opinions when they can do what they want?

**Reply by meichthys** [reply]
@eliandoran I'm not opposed to this 'faster' approach at all - especially if waiting for the review is holding anything up.
As we discussed in our first meeting, we should probably get ourselves 'out there' (on reddit, and maybe hacker news, or others?). 
Perhaps that could draw in some other interested developers.

**Reply by zerebos** [reply]
> For some PRs this has caused me to merge some PRs without review, since they were needed for the upcoming release or no consensus was reached.

Sure if that's what you want to do, I certainly don't have the time nor energy these days to move at the same pace you do.

> Regarding "as the changes are something everyone is on the same page", I think we'd have a hard time aligning everyone on everything. 

Agreed, I meant more at a very high level. I'm sure you and others wouldn't appreciate if I suddenly merged a PR of me converting the entire backend to Go without consulting anyone.

> And then comes the philosophical question: if there's only a handful of developers, why should they cater to everyone's opinions when they can do what they want?

True enough, with open source licenses I (and other developers) can simply fork and do things as we like and at our own paces. But I'm really not sure what you mean by that here specifically.

**Reply by eliandoran** [reply]
> I'm not opposed to this 'faster' approach at all - especially if waiting for the review is holding anything up.

@meichthys , I appreciate it. I think we should proceed with it for the time being.
And you're right, we can start marketing ourselves a bit. I saw Trilium was losing some traction due to it entering maintenance, for example the fact that it was/is going to be removed from Arch's AUR due to it using an older Electron.

Let's try to find a milestone for the first fully public release. I have a bit of confidence that things are going to work in the right direction. But in order to be able distinguish ourselves from upstream, we need to bring something to the table. So far we only have a refactoring (which is not a good sell for users, only for developers) and a bit of updates. Let's try to push for something else as well.

> I'm sure you and others wouldn't appreciate if I suddenly merged a PR of me converting the entire backend to Go without consulting anyone.

If performance is up by 1000x, I wouldn't mind it, haha. Maybe go for Rust instead of Go. But yeah you're right about very top-level stuff. 

> True enough, with open source licenses I (and other developers) can simply fork and do things as we like and at our own paces. But I'm really not sure what you mean by that here specifically.

I was referring to the fact that since the developers are the ones implementing the features, they should not necessarily feel constrained by the rest of the org due to lack of consensus. If the developers (like everyone else, of course) are doing it as a hobby, then why should they have to cater to everyone's decision-making? It's more the idea of meritocracy.

But regardless, since there are multiple teams involved, we still need a lead to give a general direction of the application. I think a lot of time is spent on the org _discussing_, rather than _getting things done_.

**Reply by meichthys** [reply]
I think a new logo and some slight UI improvements that you have been discussing (perhaps to the mobile interface too) would be enough for the first public release. I've seen 'next-gen' releases of many projects that have been nothing more than a name change and some new marketing flare.

**Reply by JYC333** [reply]
It's my first time joining an open source project, and my feeling is that someone need to push a bit harder to get things done.

> Regarding "as the changes are something everyone is on the same page", I think we'd have a hard time aligning everyone on everything. And then comes the philosophical question: if there's only a handful of developers, why should they cater to everyone's opinions when they can do what they want?

And I also agree on this, as long as the main direction is aligned, technical details shouldn't be too much important if there are no breaking changes. I may not have time to develop features recently, but I'll try my best to help reviewing and unblocking for the development @eliandoran . I believe the free time and the willness of putting time on this project for everyone is varying alone the time. As long as some of us could unblock the development, I think it's better to get things done before the developers losing their passions.

**Reply by zerebos** [reply]
> Maybe go for Rust instead of Go.

Can I ask why? I find Go to be much nicer to work with and Rust to be overkill for something that isn't an embedded system.

> But regardless, since there are multiple teams involved, we still need a lead to give a general direction of the application. I think a lot of time is spent on the org discussing, rather than getting things done.

Yeah this was kinda more my point, someone making the high level calls is great and the "being on the same page" should be done at the same high level. Like I shouldn't be able to just stop your work because of my personal preference, but if I bring forward a good reason to not go in a certain direction then I think that should be discussed.

**Reply by eliandoran** [reply]
@meichthys ,

Then we should get started. 💪🏻 

Either way, we need to resume the discussion about the logos. I'll start a new poll.

@JYC333 , thanks.

> Can I ask why? I find Go to be much nicer to work with and Rust to be overkill for something that isn't an embedded system.

Honestly I can't say I have much experience with Go, I contributed really small fixes to a few projects and I can't say I'm a fan of the syntax. I worked only a bit with Rust but I really appreciated the strictness of the compiler. Sometimes I feel that TypeScript is not strict enough, especially in regards with exception checking.

My "final boss" would be adding full support to mobile (that is not just client, but also server). I feel that if we are going to do it in Rust or Go it's going to be more difficult.

> Yeah this was kinda more my point, someone making the high level calls is great and the "being on the same page" should be done at the same high level

I'm glad that we're aligned on this.

**Reply by zerebos** [reply]
> and I can't say I'm a fan of the syntax.

Fair and oddly enough I feel the same way about Rust. Let's settle on Zig instead 😉 

> My "final boss" would be adding full support to mobile (that is not just client, but also server). I feel that if we are going to do it in Rust or Go it's going to be more difficult.

Time to move everything to dart/flutter 😄
